### ISSUE TITLE
Autorenewal mechanisms

### ISSUE THREAD

**Issue by schoen** [letsencrypt/letsencrypt#257]
(This would be great to get @fmarier's and @pde's opinions on, among others!  And @paravoid, if you're still following this project.)

We need to figure out mechanisms for renewing certificates automatically, if a user wants to do that. This will require creating a cron job and also keeping track of what certificates should (if any) should be autorenewed.  This might interact a bit with operating system policies about the right way to create and manage automated recurring processes.

My current thought for Debian-like systems is this:

* When we originally install, we should create a `/etc/cron.monthly/letsencrypt-renewer` script.
* If the package is uninstalled, that script will be deleted.
* The script will read a file called `/etc/letsencrypt.conf` or `/etc/letsencrypt-renewer.conf` or similar. The default action will be to do nothing and exit immediately. (So, no actual autorenewal attempts will take place merely as a result of having the package installed.)
* If particular keys and certificates are identified in that file and a directive is enabled there to try to autorenew them, the script will attempt to do so (at an appropriate time prior to its expiry), and then try to deploy the resulting certificate.
* At the time that a cert is originally issued, the user will be asked whether or not to enable automated renewal of that cert.
* The user can manually edit the config file to change the policy with respect to any individual cert or with respect to this mechanism as a while.

How does this sound?  Here are some additional questions:

* Should we regenerate keys, or only certs?
* How to notify the user when renewal was successful or unsuccessful?
* How long before expiry to try to renew?
* Do we need to explicitly check an external time source when deciding whether to make a renewal attempt, or can we believe the system clock?
* How to structure the config file? What information belongs in it?
* How to decide what installer to use in order to try to automatically deploy the reissued cert?  What kind of API does the installer need for this?
* Is it OK to spontaneously send a web server a graceful restart request in response to the deployment of a new cert, without prior human confirmation?
* How to get the user to understand the possible need to back up key material or migrate it when moving a domain or server to new hardware? (Can we also provide a higher-security migration mechanism for people who are concerned about moving key material off of the system where it was generated?)
* How does this work on non-Debian-like systems?

**Comment by fmarier** [comment]
Your plan sounds good to me. Here's a minor comment:

> The script will read a file called /etc/letsencrypt.conf or /etc/letsencrypt-renewer.conf or similar.

You could set an AUTO_RENEW variable in `/etc/default/letsencrypt` which would be sourced from the cronjob.

And here are my thoughts on some of your other questions:

> Should we regenerate keys, or only certs?

I'd personally go with keys because it's cheap and it helps in case the private key was compromised and the admin never noticed.

> How to notify the user when renewal was successful or unsuccessful?

A typical way would be to email `root`. Also, the output of the cronjob should be emailed to root automatically.

> How long before expiry to try to renew?

I'd say, at least a month, in case there are problems.

> Do we need to explicitly check an external time source when deciding whether to make a renewal attempt, or can we believe the system clock?

If the system clock is off by more than a day, there's a serious problem on that box, so that shouldn't matter for this purpose.

> How to structure the config file? What information belongs in it?

The minimum I can think of is a simple bash fragment that lists the domains to auto-renew certs for:

    #AUTO_RENEW=""
    AUTO_RENEW="login.example.com app.example.com"

> Is it OK to spontaneously send a web server a graceful restart request in response to the deployment of a new cert, without prior human confirmation?

Maybe we could start by mailing the admin:

    "A new cert for login.example.com has been issued and is ready to be installed.
    
    To do so, run this command:
    
      letsencrypt renew login.example.com
    
    If you don't do this before 28 Feb 2015, it will be done automatically.
    
    Thanks for using Let's Encrypt!"

and then do it automatically for them after a few days. Just a thought.

**Comment by schoen** [comment]
A piece of good news is that CentOS (and presumably RHEL) has `run-parts` and hence `/etc/cron.daily`, etc.

https://www.centos.org/docs/5/html/Deployment_Guide-en-US/ch-autotasks.html

**Comment by schoen** [comment]
Another important challenge is how to keep track of what the renewal process has actually done (or attempted to do). For example, it probably needs to be able to record whether an attempted renewal failed (or for that matter succeeded), to distinguish this from the case where `cron` simply failed to run the renewal script altogether, or where the machine wasn't connected to the Internet at the moment that the renewal was supposed to be attempted.

**Comment by pde** [comment]
On rekeying:

>> Should we regenerate keys, or only certs?

> I'd personally go with keys because it's cheap and it helps in case the private key was compromised and the admin never noticed.

This seems to me to be an easy example of something the config file should decide, with the default being auto key rotation.

We should in parallel offer a <tt>lets-encrypt rekey</tt> command the admin can run if they have a reason.

**Comment by pde** [comment]
Other thoughts:

- Don't ask the user for permission to auto-renew.  Tell them it's happening,† give them command line flags and config files to control or disable it if they want, but by default we should do the right thing for making cert management invisible and automatic.
- The cron job should probably run daily.  Eventually we'll want to support prompt server-initiated re-issuance too.
- Renewal should be at some percentage of cert lifetime,  like two-thirds or 80%.

† This and a couple of other things should be organised into a "post successful run" information screen that the admin gets, rather than interrupting the flow of the UI.

**Comment by pde** [comment]
On clock errors:

>> Do we need to explicitly check an external time source when deciding whether to make a renewal attempt, or can we believe the system clock?

> If the system clock is off by more than a day, there's a serious problem on that box, so that shouldn't matter for this purpose.

Unfortunately you can't trust the system clock or assume that tools like ntpdate and ntpd are present for handling time problems.  We would probably save a few people a few cert warnings and HSTS lockouts by trying to detect really wonky clocks, but it's definitely not a highest-priority feature.

I can probably revise the [ancient Switzerland time-wrangling code](http://sourceforge.net/p/switzerland/code/HEAD/tree/trunk/switzerland/client/TimeManager.py) to give you some code that makes a best-effort attempt to tell you if the system definitely has a really wonky clock.  If it's wrong by very large amounts, email the admin but also renew earlier.

**Comment by pde** [comment]
More answers:

- whatever Installer was used at first run, should also be the one that's called with the new cert / keys
- for the Standalone Authenticator, you probably don't want to do autorenewal by default.  The admin can ask for it and set up a drop point for the files if they want
- if renewals are failing, email <tt>root@localhost</tt>.  If they're really failing, and the cert is about to expire, but the cron job can still talk to the ACME server, I've thought about whether it would be a good idea to try to email the RecoveryEmail if we have one.  Maybe one day, but probably not a launch feature.

Some considerations for the config file structure, without any definite answers:

 - Like configs for ssh and a bunch of other things, it's going to have a bunch of variables, and an idea that they can be set globally or in cert-specific stanzas.
 - You could make your config file with [ConfigParser](https://docs.python.org/2/library/configparser.html) or with raw .py files like mailman does.  I'd lean toward the former unless research indicates problems with that.
 - It's going to need to identify the certs somehow.  One obvious key is path, another is sha256sum, another is serial number.  Some [other facts about the cert](https://github.com/letsencrypt/acme-spec/blob/master/draft-barnes-acme.md#proof-of-possession-of-a-prior-key) could be dropped in as helpful comments.
 - the cron job may want to remove entries in the config file for old, obsolete certs.  That's a weak argument for using a directory structure with multiple files, though I personally find those config structures a bit more work to understand.

**Comment by fmarier** [comment]
> Don't ask the user for permission to auto-renew. Tell them it's happening,† give them command line flags and config files to control or disable it if they want, but by default we should do the right thing for making cert management invisible and automatic.

I like the automatic default, but I suspect many admins will want to manually test their server to make sure that everything is working. Which is why I suggested the "apply the change manually by Friday or I'll do it for you then". In other words, fully-automatic cert renewal but with advance notice.

**Comment by schoen** [comment]
Next thoughts:

* Should there be some mechanism for grouping renewals and/or reinstallations to happen simultaneously? (If we had 6 certs on the same machine which all expire at the same time or all expire soon, do we try to detect that somehow and renew them as part of the same action? And deploy the resulting cert as part of the same action?)

* Do configurators somehow specify a particular target instance of the thing that they control? If we have 3 copies of Apache with different config files and listening on different ports, can we remember which one that Apache configurator originally used and tell the configurator to use that one for renewals too?

* Should we have a separate config file section for each individual cert under our management? The ConfigParser wants each directive (i.e., variable) in a section to have a unique name, so we can't say something like

```
[certs]
configurator: apache
privkey: /etc/letsencrypt/key-0001.pem
cert: /etc/letsencrypt/cert-0001.pem
expiry: 2015-10-10
configurator: apache
privkey: /etc/letsencrypt/key-0002.pem
cert: /etc/letsencrypt/cert-0002.pem
expiry: 2015-10-11
configurator: apache
privkey: /etc/letsencrypt/key-0003.pem
cert: /etc/letsencrypt/cert-0003.pem
expiry: 2015-10-10
```

because the last definitions of `configurator`, `privkey`, etc., would supersede and shadow all of the earlier ones.

We could conceivably have something like

```
[cert1]
configurator: apache
privkey: /etc/letsencrypt/key-0001.pem
cert: /etc/letsencrypt/cert-0001.pem
expiry: 2015-10-10

[cert2]
configurator: apache
privkey: /etc/letsencrypt/key-0002.pem
cert: /etc/letsencrypt/cert-0002.pem
expiry: 2015-10-11
```

though I wonder if users looking at the file would find it confusing if some keys or certs were deleted or removed from letsencrypt management so that we had non-contiguous ```cert3```, ```cert8```, or similar stanzas.

**Comment by jdkasten** [comment]
I didn't realize how closely tied to revocation this problem is...

I don't believe we should expect that the certs/keys that are stored for users are still going to be there or exist when renewal comes up.  I believe power users will likely rename certs/move them around.  The Revoker maintains a backup of every ACME generated certificate/authorized key based on the ACME server that issued it.  (It also maintains original paths... Revoker.Cert contains facilities to tell if the original certificate/key has changed/moved/deleted.)

When displaying all of the certs for revocation purposes... it finds out which certs are currently installed by using the selected installer's get_certs_keys function.
https://github.com/letsencrypt/lets-encrypt-preview/blob/master/letsencrypt/client/interfaces.py#L180

This returns all of the paths and then I check it against fingerprints of the backup certs to actually make the associated connections.

I will try to think about this problem a bit more, but we will want to have a plan for file movement.

Nitpick... It might make more sense to define the class that supports the interface.
```installer: ApacheConfigurator``` rather than ```configurator: apache```.  I also assume privkey means authorized key in this context.

**Comment by schoen** [comment]
How is the client going to deal with generating and tracking authorized keys? Is there going to by default be one authorized key per machine? Per private key?

**Comment by schoen** [comment]
I'm also thinking it might be better to have something with more hierarchical structure than ConfigParser provides, but human beings might be reluctant to edit JSON or XML files (or might be more likely to introduce syntax errors into them). The structure could be nice if we had many certs to renew, like

```
policies:
    autorenew: yes
certs:
    cert1:
        whentorenew: 2015-09-10
        details: [...]
        status: [...]
    cert2:
        whentorenew: 2015-09-11
        details: [...]
        status: [...]
```

Well, I guess the trouble is that for some purposes we want to think of this as a config file (letting the human sysadmin change the policy details) and for other purposes it's almost more like a database (keeping track of many details about what the cert is, when it expires, when or whether it should be renewed, how it should be installed after it's renewed, how it should be renewed (using what authorized key), the exact locations of constituent files, what happened when we last tried to renew it, whether we've already renewed it or not...).

At least the file-location parts of that might be manually edited by a human sysadmin.

**Comment by schoen** [comment]
I can make a list of specific details that should presumably be tracked about a particular cert's renewal.

**Comment by schoen** [comment]
I've been thinking about this more and sketching out some ideas.

Right now I'm thinking that there needs to be a database of certs that we know about, perhaps structured as JSON, including details about the cert content, expiry, our renewal attempts, and where on the disk the associated files can be found.  Then there should be a separate config file, perhaps in the ConfigParser format, that describes policies for autorenewal, notification, and automated installation of renewed certs.

For the database of certs, the fields I've thought of are

* subject name(s)
* internal reference name? (primarily for successor relationships)
* external reference name (serial or hash)? (for protocol use?)
* where physically stored: cert, privkey, chain
* when expiring (ISO format or Unix time!?)
* where installed and how (which configurator was used)
* renewal state:
  * when user notified
  * when renewed
  * new pending certs?
  * when installed?
  * designated successor cert upon renewal?  (because maybe we should track both of them)

The successor cert idea is that if a cert has been renewed but not yet installed, both the old and new certs exist on the system at the same time, presumably also with overlapping validity windows, so it appears that there needs to be a way to designate the relationship between them (to say that cert B is the successor of cert A, and ought to be installed in place of cert A whenever the installation event happens).

For the renewal configuration:

* autorenew on/off
* threshold for short-lived certs
* ratio for short-lived certs
* renewal window for long-lived certs
* renew notify on/off (and when)
* autoinstall on/off
* autoinstall notify on/off and when
* notify whom

**Comment by schoen** [comment]
And I have some further conceptual questions that have come up as the result of sketching out further steps in this process:

* For autorenewal, do we automatically believe or assume that each DNS name in question still refers to us?  What do we do if some or all DNS names have expired or been pointed elsewhere?  (We will also not be able to do automated renewal if the cert has been deployed on a proxy, load balancer, or CDN that is currently in between the Internet and us.)
* Can we somehow distinguish the case of "you no longer own this domain" from the case of "this domain is not currently pointed at this machine"?
* Should we renew some names but not all if the renewal validation can't complete for all the names?  Presumably we should avoid or delay automated deployment of the new cert in that case?

* How do we decide (or know) whether we can or should use particular authorized keys and where to find those keys? Is that another database field?

* Do we look at the live system (somehow) to determine whether a successor cert has been installed, or only look at our own database? What if the user manually installs the successor cert by editing the config file, rather than by running our scripts?

* What happens if the user has switched server software over the course of the lifetime of the cert? Are we going to, for example, use the Apache configurator to try to obtain and install a new cert, when in fact the user is now using Nginx (but was running Apache when the cert was originally obtained)?

**Comment by schoen** [comment]
And we'll also have a file write locking requirement, maybe, for updates to the database file (to avoid having two inconsistent updates happen at the same time).

**Comment by pde** [comment]
> For autorenewal, do we automatically believe or assume that each DNS name in question still refers to us? What do we do if some or all DNS names have expired or been pointed elsewhere? (We will also not be able to do automated renewal if the cert has been  deployed on a proxy, load balancer, or CDN that is currently in between the Internet and us.)

This question is somewhat tied to @jsha 's draft plan for which challenges will be issued, and when, by the CA.  My feeling is that we should aim to ensure that the webserver can basically always autorenew either with its Authorized + Subject keys if that's sane from a security perspective.  Re-performing DV creates a set of new opportunities for breakage during the renewal process, and we should minimise those if we can.  

So my first hunch on this is: the way you should be able to tell that you don't own a domain anymore is that the cert has been revoked, and one criterion for server behaviour is that if the WHOIS for a domain changes, it should schedule a re-run of DV, at the very least on the next renewal event, but quite possibly in some server-initiated situation that's sooner than that.

**Comment by pde** [comment]
> Should we renew some names but not all if the renewal validation can't complete for all the names? Presumably we should avoid or delay automated deployment of the new cert in that case?

There are several variables that affect this calculation: is this a SAN cert for all of the names, or a bunch of SNI certs for each one separately, or both in a bundle?  Do the webserver config and/or logs indicate that the server is still answering queries for the name that cannot be renewed?  @jdkasten would we need to add extra API hooks to export this kind of analysis from the Configurators?  

In any case all of this supports @fmarier 's suggestion that we email the admin saying "hey, we've renewed your cert and we're going to deploy it".  Or in this case we email the admin saying something that's proportional to how serious the issue appears to be.  If we still see log entries for the name we can't renew "URGENT: Cert renewal failure for <domain>; if don't see log entries or a config then the message is much more "NOTE: certs not renewed for <domain>".

**Comment by pde** [comment]
Thinking about the config file format, my instinct is that we should design it in the way that is absolutely, minimally confusing to the sysadmin.  There are a few ways to do this, but let's call this one "Plan A":
```
policies:
    autorenew: yes
certs:
    /etc/lets-encrypt/certs/example1.com.pem:
        whentorenew: 2015-09-10
        domains: example1.com www.example1.com mail.example1.com
        details: [...]
        status: [...]
    /etc/lets-encrypt/certs/example2.pem:
        whentorenew: 2015-09-10
        domains: example1.com www.example1.com mail.example1.com
        details: [...]
        status: [...]
    /etc/lets-encrypt/certs/all-domains-san.pem:
        domains: example1.com www.example1.com mail.example1.com example2.com
        whentorenew: 2015-09-11
        details: [...]
        status: [...]
```
The above option would require both the client, and perhaps a README.BEFORE.MOVING.txt file in /etc/lets-encrypt/certs to warn the admin to edit the renewal config if they want to move any files.

In the above scheme, there should probably be a directory like /etc/lets-encrypt/certs/store, which contains all certs that we've ever had; /etc/lets-encrypt/certs contains symlinks to the store.  The deployment method for the renewed certs is updating the symlinks.  We probably want to have a "lets-encrypt deploy-renewed" or similar command to that for you automatically, which we can mention in those emails to admins.

Not sure if the symlinks would work, or we'd need a different plan, on Windows.  But we can probably solve that later.

**Comment by pde** [comment]
And here's a "Plan B", using essentially nominal labels as the config keys:
```
policies:
    autorenew: yes
    renew-before-expiry: 3 months   # or 75% ?
certs:
    example1.com
        latest-filename: /etc/lets-encrypt/certs/example1.com-2015-02-01.pem
        domain: example1.com
        alternative-domains: www.example1.com mail.example1.com
        key: /etc/lets-encrypt/keys/rsa-key-eg.pem
        details: [...]
        status: [...]
    example1.com-b
        latest-filename: /etc/lets-encrypt/certs/example1.com-2015-02-01.pem
        domain: example1.com
        alternative-domains: www.example1.com mail.example1.com
        key: /etc/lets-encrypt/keys/ecc-key-eg.pem
        details: [...]
        status: [...]
```
Note that it is _almost_ possible to use the section keys as the domain name, but the fact that people might want to have two different certs for example.com messes that up if ConfigParser doesn't support repeated keys.

**Comment by jdkasten** [comment]
Remember that we may have access to a renewal URI that would not require another DV authentication.  This should certainly be stored and solves many of the issues above.

> Do the webserver config and/or logs indicate that the server is still answering queries for the name that cannot be renewed? @jdkasten would we need to add extra API hooks to export this kind of analysis from the Configurators?

@pde, Yes, that would require another API hook and I believe the information would be prone to error. (It could probably be done in a reasonably safe manner, but would require a fair amount of additional work and it would never be 100% accurate)

I am still very hesitant to rely strictly on the path name for the certificate.

1.  By default, we use very generic names that will often be changed.
2.  Great naming isn't trivial with SANs. (I imagine most certs will at least contain the www SAN.)
3.  When I am renewing certificates for my own webservers, I have a tendency to rename the certs/keys as I get the configuration the way that I want it. 

I think that the fingerprint or relevant info should be checked before acting on the information.

> How do we decide (or know) whether we can or should use particular authorized keys and where to find those keys? Is that another database field?

This info is currently stored by the revoker, perhaps a better API should be created to query this info.  Ultimately though, it is the client's responsibility to handle the best way to get authorizations/which keys are authorized for which domains.  Eventually when you run the client, it won't even solve challenges for names for which it already has authorization.  
This isn't an issue with the renewal URI.

> Do we look at the live system (somehow) to determine whether a successor cert has been installed, or only look at our own database? What if the user manually installs the successor cert by editing the config file, rather than by running our scripts?

This is expected and has been the assumption throughout development.  You need to verify that the cert you are replacing is the one you expect.  The function get_all_certs_keys exists within installers to get relevant info.  The revoker also contains copies of all the valid ACME certificates along with their original paths, and code to tell whether the path has changed, etc...

**Comment by jsha** [comment]
Regarding availability of auto-renewal without performing additional DV: I think we should require fresh DV at least once a year. I believe this brings us to parity with the level of verification required by other CAs, but I'm open to hearing about counter-examples.

In general I've been assuming a given domain validation is usable for issuance for 30 days. Ideally any renewal occurring after that 30 days would require fresh DV.

**Comment by schoen** [comment]
I guess a related further question is whether the client *can tell* ahead of time whether a renewal attempt would succeed or fail. Should a model of these policies be built into the client, or should there be an API mechanism where the client can ask "is this credential still able to cause issuance without more?"?

(Obviously the CA is still entitled to refuse to renew certs if it has some reason to do so, regardless of whether the client anticipates that outcome understands why. But I wonder whether we can usually avoid trying to cause renewals which don't end up succeeding.)

**Comment by schoen** [comment]
A related issue, on checking "intent to revoke", is just filed in letsencrypt/acme-spec#70.

**Comment by pde** [comment]
Support for the <tt>letsencrypt rekey</tt> operation now has a separate ticket, #407.

**Comment by bmw** [comment]
The design work laid out in this ticket seems to have been completed. Closing for now.
